feat: parellel vdf service initialization

ernius opened the pull request:
**Describe the changes**
Adds parallel vdf initialization function [fn create_state_parallel(](https://github.com/Irys-xyz/irys/blob/8a3a48aefa6d68505d777fb28929835bd9b09d7e/crates/actors/src/vdf_service.rs#L95-L96). Parallelization is not direct as vdf steps vary per block so is not possible to know in advance how many blocks needs to be recovered from db. So a  window of blocks are parallel retrieved and added in a loop until the required number of vdf steps is reconstructed. 

Also for this is added a new table [table IrysBlockHeadersByHeight<Key = u64, Value = H256>;](https://github.com/Irys-xyz/irys/blob/8a3a48aefa6d68505d777fb28929835bd9b09d7e/crates/database/src/tables.rs#L101) indexing block hashes by height and a corresponding function to get blocks by height [pub fn block_header_by_height<T: DbTx>(](https://github.com/Irys-xyz/irys/blob/8a3a48aefa6d68505d777fb28929835bd9b09d7e/crates/database/src/database.rs#L85-L86).

This windows was manually set up to [n = 50; // number of block headers read in parallel](https://github.com/Irys-xyz/irys/blob/8a3a48aefa6d68505d777fb28929835bd9b09d7e/crates/actors/src/vdf_service.rs#L107-L108) that ended up being the best number by checking different amount 10, 50, 100, 200.

This new method performance is compared to previous sequential initialization giving worst performance, previous is arround 500ms and in parallel results give 2s. Checked a little and found Rayon is well-suited for CPU-bound tasks but not ideal for disk I/O-bound tasks because its thread pool is designed for CPU-intensive work, not blocking I/O, so for disk I/O, seems to be best suited asynchronous libraries like tokio. So I may test using just tokio::spawn to do async db I/O operations.


**Related Issue(s)**
#264
**Checklist**

- [x ] Tests have been added/updated for the changes.
- [ x] Documentation has been updated for the changes (if applicable).
- [x ] The code follows Rust's style guidelines.

ernius commented on crates/chain/tests/block_production/testing_primitives.rs:
```diff
@@ -0,0 +1,48 @@
+use crate::utils::{get_height, mine, wait_until_height};
+
+
+async fn start_node() -> IrysNodeCtx {
```
Yes, sorry cross branch contamination :)

DanMacDonald commented on crates/database/src/tables.rs:
```diff
@@ -99,6 +99,9 @@ tables! {
     table IrysBlockHeaders<Key = H256, Value = CompactIrysBlockHeader>;
 
+    /// Stores the header hashes belonging to the canonical chain, indexed by height.
+    table IrysBlockHeadersByHeight<Key = u64, Value = H256>;
```
I believe @roberts-pumpurs  has written this code in one of his feature already, you can reference it or maybe even reuse it.

ernius replied:
I think block index do not store all the blocks history and its idea is more as a mempool of last blocks, but maybe I am wrong here, anyway I assumed that was the case. Anyway for the performance issue will not affect the overall performance.

DanMacDonald replied:
You're correct - the `block_index` stores "well confirmed" blocks that are unlikely to be reorganized, typically 2-3 blocks behind the chain head. The blocks at the head of the chain come from the `block_tree`, which manages current forks and switches between them as confirmations change.

Creating a `height`->`block_hash` table would require updates whenever a new fork is chosen, introducing potential synchronization issues between `block_tree` and `IrysBlockHeadersByHeight` table. These could cause difficult-to-diagnose bugs.

While such a table would simplify height-based header retrieval, it adds complexity by creating another source of truth for chain state. It's better to retrieve current state directly from the components that manage that state rather than introducing an additional table/index.

If this continues to be an issue, it may be a place we want to explore an abstraction, a wrapper class that is backed by the `block_index` and `block_tree` that exposes a method to retrieve block_hash by height.

ernius replied:
Thanks for the clear explanation, understood, then I will remove this new table and use block index checking @roberts-pumpurs implementation.

roberts-pumpurs replied:
[This](https://github.com/Irys-xyz/irys/blob/857c5ef65ef2abf9019a388a34e6d265517ed67e/crates/actors/src/ema_service.rs#L404-L419) is probably the module that you'd be interested in @ernius

ernius commented on crates/actors/src/vdf_service.rs:
```diff
@@ -94,6 +94,59 @@ fn create_state(
+
+    while steps_remaining > 0 && height > 0 {
+        // get in parallel n blocks
+        info!("reading from {} to {}", height.saturating_sub(n), height);
+        let blocks: Vec<IrysBlockHeader> = (height.saturating_sub(n)..height)
+            .into_par_iter()
+            .map(|h| {
+                db.view_eyre(|tx| block_header_by_height(tx, h, false))
+                    .unwrap()
+                    .unwrap()
+            })
+            .collect();
```
Good one, I will try that.

ernius commented:
Addressed comments, so removed height blocks db table, and recover block using block index to get its hash by height and with that recover the block from db. 

Also removed rayon parallelization that seems to be a bad choices for IO blocks, and use tokio block spawns to get in parallel a configurable amount of n blocks in parallel, 

using rayon times where around 2.6 seconds

VdfService vdf steps initialization time in parallel: 2.63s

and using tokio block spawns when down a second

VdfService vdf steps initialization time in parallel: 1.53s

however still the original secuential block retrival has better times
VdfService vdf steps initialization time: 485.78ms
